### PR TITLE
Fix: Can no longer slice csr_matrix in 1.3.0 

### LINKED ISSUE #11200: Can no longer slice csr_matrix in 1.3.0
I have a code example that works in scipy 1.1.0, 1.2.0, 1.2.1 and has a stack trace in 1.3.0, 1.3.1, 1.3.2.

#### Reproducing code example:

```
import pandas as pd
from scipy.sparse import csr_matrix

df = pd.DataFrame({'thing': [1, 1, 2, 2, 2],
                   'score': [0.12, 0.13, 0.14, 0.15, 0.17]})

row_index = df['thing'] == 1
print(type(row_index), row_index)
print(df[row_index])
sdf = csr_matrix(df)
print(sdf[row_index])
```

#### Error message:

```
Traceback (most recent call last):
  File "./foo.py", line 13, in <module>
    print(sdf[row_index])
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/_index.py", line 59, in __getitem__
    return self._get_arrayXslice(row, col)
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/csr.py", line 325, in _get_arrayXslice
    return self._major_index_fancy(row)._get_submatrix(minor=col)
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 690, in _major_index_fancy
    np.cumsum(row_nnz[idx], out=res_indptr[1:])
  File "<__array_function__ internals>", line 6, in cumsum
  File "root/.venv/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2423, in cumsum
    return _wrapfunc(a, 'cumsum', axis=axis, dtype=dtype, out=out)
  File "root/.venv/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 61, in _wrapfunc
    return bound(*args, **kwds)
ValueError: provided out is the wrong size for the reduction
```

#### Output of working version (on earlier versions of scipy):

```
<class 'pandas.core.series.Series'> 0     True
1     True
2    False
3    False
4    False
Name: thing, dtype: bool
   thing  score
0      1   0.12
1      1   0.13
  (0, 0)	1.0
  (0, 1)	0.13
  (1, 0)	1.0
  (1, 1)	0.13
  (2, 0)	1.0
  (2, 1)	0.12
  (3, 0)	1.0
  (3, 1)	0.12
  (4, 0)	1.0
  (4, 1)	0.12
```

#### Scipy/Numpy/Python version information:
```
1.3.0 1.17.3 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

See the same info on stackoverflow [here](https://stackoverflow.com/questions/59273855/how-to-select-only-some-rows-of-a-scipy-sparse-csr-matrix).

I'm running in OS X, tried pandas 0.25 and 0.25.3, though I don't know that those matter.

### PR DESCRIPTION
**Pull request by ananyashreyjain**
Fixes
#11200

This PR will provide a certain degree of leniency in array types used for storing boolean indices when converting them to integers. For example, after this PR, CSR matrices can be sliced directly with boolean indices even if the boolean indices are not stored as `numpy.ndarray`.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
It looks like our documentation checking code (refguide_check) is more sensitive to this change than our unit test suite, which maybe suggests we should try to have a unit test of some sort tailored to this fix (without using i.e., pandas).

It looks like some array typing has changed now?

The 32-bit Linux test failures are a bit strange---could they be indirectly related?

**Comment by tylerjereddy** [comment]
Ah, I see the sparse failures are actually *separate* from this sparse change, possibly due to NumPy 1.18: https://github.com/scipy/scipy/issues/11266

Confusing!

**Comment by ananyashreyjain** [comment]
@tylerjereddy, even I think that the failures are separate from this change, originating due to Numpy 1.18.0 #11270 and #11266

File: scipy/sparse/_index.py [hunk 1]
```diff
@@ -276,9 +276,9 @@ def _unpack_index(index):
             'Indexing with sparse matrices is not supported '
             'except boolean indexing where matrix and index '
             'are equal shapes.')
-    if isinstance(row, np.ndarray) and row.dtype.kind == 'b':
+    if hasattr(row, 'ndim') and row.dtype.kind == 'b':
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py]: Leave a comment here to explain why we're checking for `ndim` despite not accessing it. Or better, define a small helper function `_is_boolean_index` that performs these checks.

File: scipy/sparse/_index.py [hunk 2]
```diff
@@ -325,4 +325,4 @@ def _check_ellipsis(index):
 def _boolean_index_to_array(idx):
     if idx.ndim > 1:
         raise IndexError('invalid index shape')
-    return idx.nonzero()[0]
+    return np.where(idx != 0)[0]
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py]: This should be equivalent to `np.where(idx)[0]`.

File: scipy/sparse/_index.py (hunk 1 above)
Comment (ananyashreyjain) [reply scipy/sparse/_index.py]: Making a separate helper function will be a good idea.

**Comment by ananyashreyjain** [comment]
Instead of proceeding on basis on presence of attribute `ndim` which may not cover all kinds of array types (for eg, list). I think it will be better to convert to `numpy.ndarray` (if possible) and then proceed. This will cover all kinds of array types that can be used for slicing.

File: scipy/sparse/_index.py [hunk 3]
```diff
@@ -322,6 +324,20 @@ def _check_ellipsis(index):
     return index[:first_ellipsis] + (slice(None),)*nslice + tuple(tail)
 
 
+def _compatible_boolean_index(idx):
+    """Returns a boolean index array that can be converted to
+    integer array. Returns None if no such array exists """
+    if hasattr(idx, '__iter__'):
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py]: This seems a bit too permissive. For instance, a dictionary with boolean keys would be an accepted index.

How about:

```python
if not hasattr(idx, 'ndim') or not hasattr(idx, 'dtype'):
    idx = np.asanyarray(idx)
if idx.dtype.kind == 'b':
    return idx
return None
```

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

File: scipy/sparse/_index.py (hunk 3 above)
Comment (ananyashreyjain) [reply scipy/sparse/_index.py]: Yes, this seems to simpler and better. I will make the changes.

**Review by perimosocordiae — APPROVED** [review]
Looks good. We may encounter a performance hit when speculatively calling `np.asanyarray` in cases where the element type isn't boolean, but I think the extra compatibility is worth it here.

**Comment by perimosocordiae** [comment]
@ananyashreyjain could you add a test case that indexes a sparse matrix with a list of booleans, and checks that it's equivalent to the same thing using an array of booleans?

**Comment by ananyashreyjain** [comment]
@perimosocordiae I have added some checks to avoid the performance hit for the case with non-boolean element types. I have also added a test case to ensure that indexing with boolean list works fine. The failure in Travis CI doesn't look related to this PR.

**Comment by ananyashreyjain** [comment]
Okay, this won't work with 2D Index list. I will change it back.

**Comment by ananyashreyjain** [comment]
Okay, this won't work with 2D array types of boolean indices. Some more modifications will be needed to make it work for 2D array types.

**Comment by ananyashreyjain** [comment]
This should now be compatible with all array types single or double dimensional without any performance hit for the case with non-boolean element types.

File: scipy/sparse/_index.py [hunk 4]
```diff
@@ -266,8 +266,13 @@ def _unpack_index(index):
...
             raise IndexError('invalid number of indices')
     else:
-        row, col = index, slice(None)
-
+        idx = _compatible_boolean_index(index)
+        if idx is None:
+            row, col = index, slice(None)
+        elif idx.ndim < 2:
+            return _boolean_index_to_array(idx), slice(None)
+        elif idx.ndim == 2:
```
Comment (ananyashreyjain) [inline scipy/sparse/_index.py]: Earlier `col` was set to `None` for non `numpy.ndarray`s  irrespective of the dimension which is why the 2D index array couldn't have worked.

File: scipy/sparse/_index.py [hunk 5]
```diff
@@ -266,8 +266,13 @@ def _unpack_index(index):
         else:
             raise IndexError('invalid number of indices')
     else:
-        row, col = index, slice(None)
```
Comment (ananyashreyjain) [inline scipy/sparse/_index.py]: Earlier `col` was set to `None` for non `numpy.ndarray`  irrespective of the dimension, which is why the 2D index array couldn't have worked.

File: scipy/sparse/_index.py [hunk 6]
```diff
@@ -322,7 +329,33 @@ def _check_ellipsis(index):
...
+            # Checks if index array is 2D
+            elif hasattr(first, '__iter__'):
+                ff = next(iter(first), None)
+                if isinstance(ff, bool):
+                    idx = np.asanyarray(idx)
+                    if idx.dtype.kind == 'b':
+                        return idx
+    elif idx.dtype.kind == 'b':
+        return idx
+    return None
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py]: This is a lot of nesting as written. How about:

```python
def _maybe_bool_ndarray(idx):
    idx = np.asanyarray(idx)
    if idx.dtype.kind == 'b':
        return idx
    return None

def _first_element_bool(idx):
    try:
        first = next(iter(idx), None)
    except TypeError:
        return None
    if isinstance(first, bool):
        return True
    return _first_element_bool(first)

def _compatible_boolean_index(idx):
    if hasattr(idx, 'ndim') or _first_element_bool(idx):
        return _maybe_bool_ndarray(idx)
    return None
```

**Review by perimosocordiae — APPROVED** [review]
I left a style comment, but overall this is ready to merge. Thanks for sticking with it, @ananyashreyjain !

File: scipy/sparse/_index.py (hunk 6 above)
Comment (ananyashreyjain) [reply scipy/sparse/_index.py]: Thanks, I have made the changes.

**Comment by ananyashreyjain** [comment]
@perimosocordiae Thanks for all your help. I have the made the style changes and I hope this PR can be merged now.

**Comment by perimosocordiae** [comment]
Merged. Thanks for sticking with this, @ananyashreyjain!

**Pull request merged**